src/backend/distributed/shared_library_init.c

```diff
@@ -2132,12 +2154,62 @@ StatisticsCollectionGucCheckHook(bool *newval, void **extra, GucSource source)
 static void
 CitusAuthHook(Port *port, int status)
 {
+	if (!IsCitusInternalBackend())
+	{
+		int maxClientConnections =
+			MaxConnections - ReservedBackends - ReservedCitusInternalBackends;
+
+		/*
+		 * Limit non-superuser client connections if citus.internal_reserved_connections
+		 * is set.
+		 */
+		if (ReservedCitusInternalBackends > 0 &&
+			!IsSuperuser(port->user_name) &&
+			GetAllActiveClientBackendCount() + 1 > maxClientConnections)
```

Comment:
There's a small race condition here where we might allow simultaneous connections to go through before they increase the counter and exceed the maximum.

Reply:
I think the race condition might cause future `IsCitusInternalBackend`s to fail, because we might be using more connections than `maxClientConnections` and leaving less for `IsCitusInternalBackend()`s.

And, my memory reminds me that with high concurrent pgbench -- especially at the initial phase of it--, this counter might be very busy and hence the race condition may have a higher chance.

Given that most of the time the clients do stay for a relatively long time, I think we can consider this phase as a `startup phase`, where we may consider some overhead. 

So, we could actually move the error check after `IncrementClientBackendCounter`
```C
	if (!IsCitusInternalBackend())
	{
	
		RegisterClientBackendCounterDecrement();
		IncrementClientBackendCounter();

	     int maxClientConnections =
		MaxConnections - ReservedBackends - ReservedCitusInternalBackends;

		/*
		 * Limit non-superuser client connections if citus.internal_reserved_connections
		 * is set.
		 */
		if (ReservedCitusInternalBackends > 0 &&
			!IsSuperuser(port->user_name) &&
			GetAllActiveClientBackendCount() >= maxClientConnections)
		{
			ereport(FATAL, (errcode(ERRCODE_TOO_MANY_CONNECTIONS),
							errmsg("remaining connection slots are reserved for "
								   "non-replication superuser connections"),
							errdetail("the server is configured to accept up to %d "
									  "regular client connections",
									  maxClientConnections)));
		}


	}
```

What we lose is the unnecessary calls to `RegisterClientBackendCounterDecrement/IncrementClientBackendCounter` when the client is going to fail. That seems OK?

We still have the small race condition, but from the other angle. In other words, we might throw unnecessary errors for some clients, but do not throw any unnecessary errors for  internal clients.

Reply:
Thanks, that way seems preferable!